spark-knn-core/src/main/scala/org/apache/spark/ml/knn/KNN.scala

```diff
@@ -228,12 +265,16 @@ class KNNModel private[ml](
   def setBufferSize(value: Double): this.type = set(bufferSize, value)
 
   //TODO: All these can benefit from DataSet API in Spark 1.6
-  override def transform(dataset: DataFrame): DataFrame = {
-    val merged = transform(dataset, topTree, subTrees)
+  override def transform(dataset: Dataset[_]): DataFrame = transform(dataset, withDistance = false)
```

Comment:
Ya.  You should feel free to add and change these package private functions if necessary. I'm thinking usage pattern would look like ` val model = knn.fit(df).setK(6).setMaxDistance(1).withDistance(true)` in the end.